### ISSUE TITLE
Snapshots don't instantly mount in the dataset's `.zfs/snapshot` directory

### ISSUE THREAD

**Issue by jmpesp** [oxidecomputer/crucible#827]
When creating a snapshot during the finalizing an image upload, the pantry had a requested snapshot:

	18:07:26.653Z INFO crucible-pantry (dropshot): accepted connection
	    local_addr = [fd00:1122:3344:102::5]:17000
	    remote_addr = [fd00:1122:3344:101::4]:35744
	18:07:26.654Z INFO crucible: flush with snap requested
	18:07:26.716Z INFO crucible-pantry (dropshot): request completed
	    local_addr = [fd00:1122:3344:102::5]:17000
	    method = POST
	    remote_addr = [fd00:1122:3344:101::4]:35744
	    req_id = 51bbc8b7-5c84-4319-a9ec-6fbb4208efa4
	    response_code = 204
	    uri = /crucible/pantry/0/volume/c0a6042a-6c60-4e69-872c-951a06778f44/snapshot

The next step of the snapshot create saga is to validate that those snapshots exist, and create running snapshots out of those. However, here's the agent saying that the snapshot doesn't exist:

	18:07:26.888Z INFO crucible-agent (datafile): checking if path "/data/regions/f7d172bf-3c56-4aa5-a0ff-8057c720ca76/.zfs/snapshot" exists
	18:07:26.888Z INFO crucible-agent (datafile): checking path "/data/regions/f7d172bf-3c56-4aa5-a0ff-8057c720ca76/.zfs/snapshot"
	18:07:26.888Z INFO crucible-agent (dropshot): request completed
	    error_message_external = Not Found
	    error_message_internal = region RegionId("f7d172bf-3c56-4aa5-a0ff-8057c720ca76") snapshot "4d8bc95c-a82d-4942-b7ff-d21a7518afa8" not found
	    local_addr = [fd00:1122:3344:102::7]:32345
	    method = GET
	    remote_addr = [fd00:1122:3344:101::4]:53379
	    req_id = 23fcdea4-dfc8-44e5-a115-33d30f2c03c9
	    response_code = 404
	    uri = /crucible/0/regions/f7d172bf-3c56-4aa5-a0ff-8057c720ca76/snapshots/4d8bc95c-a82d-4942-b7ff-d21a7518afa8

Checking later, It does though:

	root@oxz_crucible_oxp_4ac23275-d8fc-e984-b72c-a5d148ffde7c:~# ls /data/regions/f7d172bf-3c56-4aa5-a0ff-8057c720ca76/.zfs/snapshot
4d8bc95c-a82d-4942-b7ff-d21a7518afa8

This caused the snapshot-create saga to fail, and the unwind fails too because it can't clean up a snapshot that isn't there. It seems like 172ms isn't long enough to wait for a snapshot to be mounted in the dataset's directory.

**Comment by jmpesp** [comment]
One way around this to wait in the downstairs' `region_flush` until it does appear - this would maintain that if the flush completes successfully then .zfs/snapshot directory exists.

**Comment by jmpesp** [comment]
Note this issue may be explained if `zfs snapshot dataset@name` is an asynchronous request, not a blocking call.

**Comment by jclulow** [comment]
I don't think we should use the `.snapshot` directory lookup to confirm if a snapshot exists.  I'm a bit surprised it took at least 172ms to actually show up, but either way, we should use the proper management interfaces, i.e., `zfs list` or `zfs get`, to confirm that a snapshot exists.
